File: 2024/day8.py
import AOCInit
import util

from collections import defaultdict

if __name__ != '__main__':
    exit()

inp = util.getInput(d=8, y=2024)

nodeMap = inp.splitlines()
n = len(nodeMap)
m = len(nodeMap[0])

nodeDict = defaultdict(list)
for i in range(n):
    for j in range(m):
        if (c := nodeMap[i][j]) != '.':
            nodeDict[c].append((i, j))

# part 1
anodes = set()
for v in nodeDict.values():
    l = len(v)
    for i in range(l - 1):
        for j in range(i + 1, l):
            if (q1 := v[i][0] - v[j][0]) % 3 == 0 and (q2 := v[i][1] - v[j][1]) % 3 == 0:
                anodes.add((v[i][0] + q1 // 3, v[i][1] + q[1] // 3))
                anodes.add((v[j][0] - q1 // 3, v[j][1] - q[1] // 3))
            pt = (2 * v[i][0] - v[j][0], 2 * v[i][1] - v[j][1])
            if util.in2DRange(pt, n, m):
                anodes.add(pt)
            pt = (2 * v[j][0] - v[i][0], 2 * v[j][1] - v[i][1])
            if util.in2DRange(pt, n, m):
                anodes.add(pt)
print(len(anodes))

# part 2
anodes = set()
for v in nodeDict.values():
    l = len(v)
    for i in range(l - 1):
        for j in range(i + 1, l):
            delta = (v[j][0] - v[i][0], v[j][1] - v[i][1])
            g = util.gcd(*delta)
            if g == 0:
                g = max(delta)
            delta = (delta[0] // g, delta[1] // g)
            x, y = v[i]
            while True:
                if not util.in2DRange((x, y), n, m):
                    break
                anodes.add((x, y))
                x += delta[0]
                y += delta[1]
            x, y = v[i]
            while True:
                if not util.in2DRange((x, y), n, m):
                    break
                anodes.add((x, y))
                x -= delta[0]
                y -= delta[1]
print(len(anodes))

File: 2024/day8.cpp
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "..\util.h"

int main(int, char**){
    std::stringstream inpStream;
    util::getInput(2024, 8, inpStream);
    std::vector<std::string> nodeMap(util::splitline(inpStream));
    int n = nodeMap.size(), m = nodeMap[0].size();
    std::unordered_map<char, std::vector<int>> nodeDict;
    for (int i = 0; i < n; ++i){
        for (int j = 0; j < m; ++j){
            if (nodeMap[i][j] == '.'){
                continue;
            }
            nodeDict[nodeMap[i][j]].push_back(i * m + j);
        }
    }

    // part 1
    std::unordered_set<int> anodes;
    for (auto &pr : nodeDict){
        const auto &v = pr.second;
        int l = v.size();
        for  (int i = 0; i < l - 1; ++i){
            for (int j = i + 1; j < l; ++j){
                int xi = v[i] / m, xj = v[j] / m, yi = v[i] % m, yj = v[j] % m;
                if (util::in2DRange(2 * xi - xj, 2 * yi - yj, n, m)){
                    anodes.insert(2 * v[i] - v[j]);
                }
                if (util::in2DRange(2 * xj - xi, 2 * yj - yi, n, m)){
                    anodes.insert(2 * v[j] - v[i]);
                }
            }
        }
    }
    util::output(anodes.size());

    // part 2

    anodes.clear();
    for (auto &pr : nodeDict){
        const auto &v = pr.second;
        int l = v.size();
        for  (int i = 0; i < l - 1; ++i){
            for (int j = i + 1; j < l; ++j){
                anodes.insert(v[i]);
                int xi = v[i] / m, xj = v[j] / m, yi = v[i] % m, yj = v[j] % m;
                int xx = xi, yy = yi, xdelta = xj - xi, ydelta = yj - yi;
                while (util::in2DRange(xx, yy, n, m)){
                    anodes.insert(xx * m + yy);
                    xx += xdelta;
                    yy += ydelta;
                }
                xx = xi;
                yy = yi;
                while (util::in2DRange(xx, yy, n, m)){
                    anodes.insert(xx * m + yy);
                    xx -= xdelta;
                    yy -= ydelta;
                }
            }
        }
    }
    util::output(anodes.size());
    return 0;
}
